### ISSUE TITLE
Approaches to Scrape Metadata to Relational plus Refactoring Proposal

### ISSUE THREAD

**Issue by arcturustech** [rarcega/instagram-scraper#100]
I'm interested in periodically scraping selected instagram metadata for a few selected public accounts and posts into relational tables (specifically Postgres), rather than json format, for analysis.

Zero'th idea for an approach of course was to use the official API, but, yeah, that requires an approval process and it doesn't look like it's intended for anything like what I have in mind. I assume this is why instagram-scraper continues to be maintained.

First super quick-and-dirty thought was to use the instagram-scraper utility as-is to grab the necessary metadata into json files and then write a separate utility to parse those into a relational database, calling each in turn from a bash script. But, this approach seemed rather kludgy, and it's two passes over the data, with an unnecessary trip to disk in-between.

Second thought was to import InstagramScraper into my own utility, instantiate it with the params I need, call scraper.scrape(), and then loop through the saved json files with my parse and db interface code. This seems only slightly less kludgy, and it's still making two passes, with unnecessary disk storage of json between passes.

Third thought was similar except to subclass InstagramScraper and override methods as necessary to get to what I need. This is not looking like it would be that easy or clean, requiring overriding functions all over the place to add what I need.

My final thought, the most work but I believe the cleanest approach, would be to fork the instagram-scraper repo and refactor the code to natively support more flexible output options, including relational db. This would take the package from mainly a special purpose utility to more of a callable package of Instagram scraping functionality, plus a convenient commandline interface to that functionality. But, obviously if I go that route, I want to consult you guys upstream first.

So, how would you guys feel about some (fairly substantial) refactorings? Specifically
* breaking the main routine and commandline argument handling into a separate file from the class
* refactoring the big InstagramScraper class into multiple classes, including lower-level objects corresponding to Account, Post, Media, Comment, Follower, Hashtag, etc., allowing calling of just the parts needed.
* For each of these lower level classes, cleanly refactor the output/persistence functionality (the media downloading and json file saves) out from the network calls and parsing functionality, to provide multiple output options.

@bebensiganteng requested a callbacks feature, which might be yet another way to meet my need, but I believe a refactoring into smaller pieces of functionality would be a cleaner design, which might also accomplish the purpose for which callbacks were requested.

So, seeking feedback on the merits of each approach, and on whether pull requests for such refactorings would be welcome if I went that direction, as well as any other ideas or suggestions.

**Comment by rarcega** [comment]
I am in favor of the last approach; to refactor the code completely. It's something I've considered before. As you noted, the current approach has its limitations which stems from its humble origins as a simple command-line utility for scraping pictures. I welcome any changes to make it more robust and facilitate its use in other contexts such as the analysis of the metadata. 

Your breakdown of the redesign is in line with what I had in mind. In particular, the refactoring of classes, separation of responsibilities via separate modules for network requests, transformation/output, etc., as well as adding full test coverage. 

I'm willing to assist the effort in any way I can. I am open to forking or creating a new repo entirely. Let me know how you'd like to proceed.

**Comment by arcturustech** [comment]
Excellent! We seem to be very much on the same page. It seems like most obvious roadmap for the future. Let me give some further thought to how best to proceed, and what assistance I might find helpful, and get back to you soon.

**Comment by arcturustech** [comment]
Initially, I had in mind a traditional refactoring - a series of gradual changes to the existing code base to gradually re-work it toward the solid design we had in mind, keeping it functional at all times.

But, since you are amenable to starting a new repo, that opens up a whole other possibility - start from scratch with a new code base, built from the ground up on a solid architecture, as opposed to something gradually evolved over time from a simple utility. All cruft would be left behind. Implement a coding standard. Probably initialize the repo from one of cookie-cutter's python templates.

I would draw from the existing code only those bits that do the magic, which as I see it is the actual calls to instagram servers (i.e. the actual scraping). In this plan, we would put the current v1.x code base into a maintenance/bug fix mode, and work the new code base until it can duplicate the capabilities of the existing 1.x, compatible at the level of the commandline interface. That would then eventually get released as instagram-scraper v2.0 and the 1.x repo would get locked at that time.

So, if you are amenable, I would prefer this new repo/new code base approach. I think it would be much less total work and lead to a much more clean and solid end product than trying to "refactor in place".

The crown jewel of the existing code is definitely the instagram server calls. You obviously have done a whole lot of finicky reverse engineering to understand their calls that I would not want to duplicate.

As for the rest, it would be mostly a new skeleton of classes, methods and commandline code on which to hang that access code, (as well as all-new transform and output code). I would focus on replacing that skeleton, drawing the access code from the v1.x code base.

The greatest assistance you could be to me would be in your understanding of how to talk to instagram's servers and get them to give us all the ripe, juicy metadata and media file url's we need to grab, and stuff them into Python datatypes. I will reuse your instagram access code from v1.x as much as possible, but I expect I will probably need help understanding how it works and what it does, what exactly it returns - none of that is really documented. Also, I might (or might not) need additional calls reverse engineered. Basically, your expertise in accessing and scraping instagram will be absolutely invaluable.

Anyone else who has a need could begin to hang additional transformations and output formats onto that skeleton. Hmmm, now that I think about it, maybe even via a plugin architecture? Something else to think about. A plugin architecture for transforms and outputs would keep things like db and filesystem access code out of the main codebase, which is properly concerned just with getting the data out of instagram and into Python objects.)

Secondarily, I could really use your (or someone's) help with building that test infrastructure.

So, in summary, major components and responsibilities for v2.0 would be:
* Overall Class Architecture and Skeleton Code (me)
* Commandline Utility calling the class architecture (me)
* Instagram Access Code within appropriate class methods (drawn from v1.x code, with help and possible expansion from you)
* Implement Plugin Infrastructure for Transforms and Outputs (me or other volunteer)
* Individual Plugins (whoever needs a particular transform or output format)
* Test Infrastructure (you or other volunteer)
* Sphynx Documentation (volunteer needed)

Your thoughts (or anyone else's)?

**Comment by arcturustech** [comment]
Hey, looking at your profile, I just found these projects, that you apparently contributed to at some point:
https://github.com/ping/instagram_private_api
https://github.com/ping/instagram_private_api_extensions
Are these any good? Would it be worth re-architecting instagram-scraper around these libraries?

**Comment by ChrisCintron** [comment]
I would be interested in helping to re-create the testing infrastructure. Let me know what best way I can be of assistance!
